File: Go/Trie/208_Implement Trie (Prefix Tree).go
type Trie struct {
	children map[rune]*Trie
	end      bool
}

func Constructor() Trie {
	return Trie{
		children: make(map[rune]*Trie)}
}

func (this *Trie) Insert(word string) {
	for _, w := range word {
		if _, ok := this.children[w]; !ok {
			this.children[w] = &Trie{
				children: make(map[rune]*Trie),
			}
		}
		this = this.children[w]
	}
	this.end = true
}

func (this *Trie) Search(word string) bool {
	for _, w := range word {
		if _, ok := this.children[w]; !ok {
			return false
		}
		this = this.children[w]
	}
	return this.end
}

func (this *Trie) StartsWith(prefix string) bool {
	for _, p := range prefix {
		if _, ok := this.children[p]; !ok {
			return false
		}
		this = this.children[p]
	}
	return true
}

/**
 * Your Trie object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Insert(word);
 * param_2 := obj.Search(word);
 * param_3 := obj.StartsWith(prefix);
 */

File: C++/Trie/208_Implement Trie (Prefix Tree).cpp
#include<string>
struct TrieNode{
    vector<TrieNode*> children{vector<TrieNode*>(26)};
    bool end; 
};

class Trie {
    TrieNode* head;    
public:
    Trie() {
        head = new TrieNode();
    }
    
    void insert(string word) {
        TrieNode *temp = this->head;
        for (auto w : word)
        {
            int index = (int)w - 'a';
            if (temp->children[index]==nullptr)
                temp->children[index] = new TrieNode();
            temp = temp->children[index]; 
        }
        temp->end = true;
    }
    
    bool search(string word) {
        TrieNode *temp = this->head;
        if (temp == nullptr)
            return false;
        for (auto w : word)
        {
            int index = (int)w - 'a';
            if (temp->children[index] == nullptr)
                return false;
            temp = temp->children[index];
        }
        return temp->end;
    }
    
    bool startsWith(string prefix) {
        TrieNode *temp = this->head;
        if (temp == nullptr)
            return false;
        for (auto w : prefix)
        {
            int index = (int)w - 'a';
            if(temp->children[index] == nullptr)
                return false;
            temp = temp->children[index];
        }
        return true;    
    }
};


/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */